### PR TITLE
Strengthen handling of unavailable cgroup stats

### PR DESCRIPTION
**Pull request by jasontedor**
On some systems, cgroups will be available but not configured. And in
some cases, cgroups will be configured, but not for the subsystems that
we are expecting (e.g., cpu and cpuacct). This commit strengthens the
handling of cgroup stats on such systems.

Relates #21029

### CODE REVIEW

**Comment by s1monw** [comment]
should we just throw FNF and catch on a top level as we do in other cases?

**Comment by jasontedor** [comment]
> should we just throw FNF and catch on a top level as we do in other cases?

@s1monw We can throw if `/proc/self/cgroup` is not available, which we do now, but we shouldn't throw if some `/sys/fs/cgroup...` is not available because it might be the case that the `cpu` subsystem is configured but not the `cpuacct` subsystem. Doing it the way that I've done it at least lets us get what is available.

**Comment by s1monw** [comment]
@jason ok can we check this higher up to skip the entire query of this subsystem

**Comment by jasontedor** [comment]
> ok can we check this higher up to skip the entire query of this subsystem

Sorry, I wasn't completely clear in my previous comment. We do already check if the subsystem is available or not, but it might still be the case that the necessary mounts to `/sys/fs/cgroup/` are not available and this can vary between subsystems (whether or not they are configured, whether or not the mounts to `/sys/fs/cgroup` are available).

**Comment by jasontedor** [comment]
@s1monw I pushed 29ed3ba68f122c3f064e631a3f8205ce118fa570 which might be more to your tastes? :smile:

**Comment by s1monw** [comment]
looks great I would be even happier if we had a static method that says `boolean isCgroupsAvailable()` and add the empty check back?

**Comment by s1monw** [comment]
@jasontedor  I am ok with pushing this since some builds are failing but I think we can do better with a nice util method?

**Comment by jasontedor** [comment]
> I am ok with pushing this since some builds are failing but I think we can do better with a nice util method?

I pushed 4c8a2cc72d7dddd460cc28117c6d1623f77374fb.

File: core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java [hunk 1]
```diff
@@ -211,26 +212,31 @@ private String readSingleLine(final Path path) throws IOException {
...
-            // matching is lazy; this can not happen in an assert
-            // as assertions might not be enabled
-            final boolean matches = matcher.matches();
-            assert matches : line;
-            // at this point we have captured the subsystems and the
-            // control group
-            final String[] controllers = matcher.group(1).split(",");
-            for (final String controller : controllers) {
-                controllerMap.put(controller, matcher.group(2));
+        if (lines.isEmpty()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java]: isn't this obsolete now? we don't return empty form this method?

File: core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java [hunk 2]
```diff
@@ -427,23 +432,58 @@ String readSysFsCgroupCpuAcctCpuAcctCfsQuota(final String controlGroup) throws I
...
 
     /**
+     * Checks if cgroup stats are available by checking for the existence of {@code /proc/self/cgroup},
+     * {@code /sys/fs/cgroup/cpu}, and {@code /sys/fs/cgroup/cpuacct}.
+     *
+     * @return {@code true} if the stats are available, otherwise
+     * {@code false}
+     */
+    @SuppressForbidden(reason = "access /proc/self/cgroup, /sys/fs/cgroup/cpu, and /sys/fs/cgroup/cpuacct")
+    private boolean areCgroupStatsAvailable() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java]: 👍

File: core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java [hunk 3]
```diff
@@ -427,23 +432,58 @@ String readSysFsCgroupCpuAcctCpuAcctCfsQuota(final String controlGroup) throws I
...
-            final String cpuControlGroup = controllerMap.get("cpu");
-            final String cpuAcctControlGroup = controllerMap.get("cpuacct");
-            return new OsStats.Cgroup(
-                cpuAcctControlGroup,
-                getCgroupCpuAcctUsageNanos(cpuAcctControlGroup),
-                cpuControlGroup,
-                getCgroupCpuAcctCpuCfsPeriodMicros(cpuControlGroup),
-                getCGroupCpuAcctCpuCfsQuotaMicros(cpuControlGroup),
-                getCgroupCpuAcctCpuStat(cpuControlGroup));
+            if (!areCgroupStatsAvailable()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java]: 👍

File: core/src/main/java/org/elasticsearch/monitor/os/OsStats.java [hunk 4]
```diff
@@ -374,7 +374,7 @@ public Cgroup(
 
         @Override
         public void writeTo(final StreamOutput out) throws IOException {
-            out.writeString(cpuAcctControlGroup);
+            out.writeOptionalString(cpuAcctControlGroup);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/monitor/os/OsStats.java]: you require it non-null above but here it's optional?

**Review by s1monw — COMMENTED** [review]
left some questions, LGTM otherwise

File: core/src/main/java/org/elasticsearch/monitor/os/OsStats.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/monitor/os/OsStats.java]: That was a bug, I cleaned up all the optionality in 4c8a2cc72d7dddd460cc28117c6d1623f77374fb but missed this one. I pushed 1d53577c0254a37a26997459c4eb453d29011c18.

File: core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java]: I pushed 8c7d7be253f5c524c6d1a8aa0ae36af2c438401e.

**Comment by jasontedor** [comment]
Thanks @s1monw, I've responded to your feedback.

**Comment by s1monw** [comment]
LGTM 👍

**Comment by jasontedor** [comment]
Thanks @s1monw.

**Pull request merged**